### DISCUSSION TITLE
Is it possible to turn off the persistent marker animation

### DISCUSSION THREAD

**Discussion by peshkira** [patrykandpatrick/vico#563] [Questions] [answered]
Hi!

First things first - as many before me - let me commend you on the fantastic library 👏. This is arguably one of the best libraries, I've ever used.

At the moment, I am evaluating if I can use it for a project of mine and I try Vico 2.0.0.alpha6 with compose-m3.
What I am trying to achieve is display a persistent marker (or at least visually distinguish) the last value on a line chart.
The data of the chart is updated at a periodic rate, where a new value is appended.

In general, I am able to do it, but at the moment, it seems that the persistent marker always animates from 0 to the value, which looks a bit odd. Is it possible to control or disable this animation.

To display the marker, I have copied your own internal `rememberMarker` Composable. (On a side note, it would be great, if you can make this function public or provide a default implementation as part of the library).
It works, but the marker dot always starts at y=0 and animates through the guideline until it reaches the value of the new data point within the graph.

Do you have an advice for me, how I can change/disable this particular animation? Ideally, the dot should travel directly (along the line) to the new point, but if not possible, just get rendered at the new point, instead of animating from 0.

Thanks very much for your help. If you need further information, I am happy to share more.

**Comment by patrickmichalik** [comment]
Hello! Thanks for the kind words. This animation is tied to how `ColumnCartesianLayer` and `LineCartesianLayer` handle new data. When a new entry is added, its _y_ value is animated from zero. Just to confirm, would you like to turn this off too, or would you like to keep this animation but have the persistent `Marker` ignore it?

**Reply by peshkira** [reply]
Hi!

I would like to keep the default animation of the data if possible (especially if the data set is significantly changed), but not let the marker animate (the latter in your question). In other words, I'd like to have two different types of animations if possible for data and marker.

I noticed that I can change the `diffAnimationSpec` to snap() and then the marker just blinks with each update, but then the whole chart animation is snap, which unfortunately does not look as smooth as otherwise.

Would this be (eventually) possible, do you think?

**Reply by patrickmichalik** [reply]
Got it, @peshkira. You can achieve this by creating a `MarkerComponent` subclass. At the moment, several `MarkerComponent` members are private instead of protected, and as a result, you have to copy over a few functions’ definitions to make the required override. The visibility of these members will be changed in Vico 2.0.0 Alpha 7, which we plan on releasing this week, making the override significantly simpler. Would you like me to show you how to make the override in Vico 2.0.0 Alpha 6, or shall we wait until Vico 2.0.0 Alpha 7 is released? (We can also do both, of course.)

Let me quickly address the remaining points. Regarding having the indicator travel along the line, I’m afraid this isn’t possible yet. As for adding `rememberMarker` to Vico itself, simpler `Marker`s can be created via `rememberMarkerComponent`, but seeing that people often copy the `rememberMarker` function from the sample app, we’ll look into adding such a function to the library API, as suggested. Thanks for the idea!

**Reply by peshkira** [reply]
Thanks @patrickmichalik !

That sounds awesome. I can certainly wait for 2.0.0 Alpha 7 and try it on my own in the meantime. No pressure for that at the moment.

Yeah, the sample marker is a quite good default marker, I'd say. Especially the dot itself, but also the label and the vertical guideline are great. If you add these to the public API, that would be amazing.

Thanks for getting back to me. Looking forward to the new release.

**Reply by patrickmichalik** [reply]
Thanks for letting me know! Once we’ve released Vico 2.0.0 Alpha 7, I’ll share a code snippet here.

**Reply by patrickmichalik** [reply, marked as answer]
[Vico 2.0.0 Alpha 7](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.7) is now available. Since you’re using the sample app’s `rememberMarker` function, you’re already creating a `MarkerComponent` subclass. Please head over to where the subclass is created (`object : MarkerComponent`) and override `draw` as follows. This should produce the desired result.

```kt
override fun draw(
    context: DrawContext,
    bounds: RectF,
    markedEntries: List<Marker.EntryModel>,
    chartValues: ChartValues,
) {
    with(context) {
        drawGuideline(context, bounds, markedEntries)
        val halfIndicatorSize = indicatorSizeDp.half.pixels
        markedEntries.forEach { model ->
            onApplyEntryColor?.invoke(model.color)
            val y =
                bounds.bottom -
                    ((model.entry as LineCartesianLayerModel.Entry).y - chartValues.getYRange(null).minY) /
                        chartValues.getYRange(null).length * bounds.height()
            indicator.draw(
                context,
                model.location.x - halfIndicatorSize,
                y - halfIndicatorSize,
                model.location.x + halfIndicatorSize,
                y + halfIndicatorSize,
            )
        }
        drawLabel(context, bounds, markedEntries, chartValues)
    }
}
```

**Reply by peshkira** [reply]
This is awesome. Thanks! I really appreciate your time and the work you do!

I will mark this as the answer and might open other discussions soon as I am very curious about the library and other features.

**Reply by patrickmichalik** [reply]
You’re welcome, @peshkira! Feel free to reach out if you have any questions.

**Reply by peshkira** [reply]
I am copying a modified version of the response that works with 2.0.0.alpha21 and a `DefaultCartesianMarker` if anyone stumbles upon this and needs it.

```kotlin
override fun draw(context: CartesianDrawContext, targets: List<CartesianMarker.Target>) {
  with(context) {
      drawGuideline(targets)
      val halfIndicatorSize = (indicatorSizeDp / 2).pixels

      targets.filterIsInstance<LineCartesianLayerMarkerTarget>().forEach { target ->
          target.points.forEach { point ->
              val y =
                  chartBounds.bottom -
                      (point.entry.y - chartValues.getYRange(null).minY) /
                      chartValues.getYRange(null).length * chartBounds.height()
              drawIndicator(target.canvasX, y, point.color, halfIndicatorSize)
          }
      }
  }
  drawLabel(context, targets)
}
```

**Reply by patrickmichalik** [reply]
Thanks for sharing!
